Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp

```diff
@@ -17062,6 +17062,68 @@ void SpeculativeJIT::compileStringLocaleCompare(Node* node)
     strictInt32Result(resultGPR, node);
 }
 
+void SpeculativeJIT::compileStringIndexOf(Node* node)
+{
+    if (node->child3()) {
+        SpeculateCellOperand base(this, node->child1());
+        SpeculateCellOperand argument(this, node->child2());
+        SpeculateInt32Operand index(this, node->child2());
+
+        GPRReg baseGPR = base.gpr();
+        GPRReg argumentGPR = argument.gpr();
+        GPRReg indexGPR = index.gpr();
+
+        speculateString(node->child1(), baseGPR);
+        speculateString(node->child2(), argumentGPR);
+
+        std::optional<UChar> character;
+        String searchString = node->child2()->tryGetString(m_graph);
+        if (!!searchString) {
+            if (searchString.length() == 1)
+                character = searchString.characterAt(0);
+        }
+
+        flushRegisters();
+        GPRFlushedCallResult result(this);
+        GPRReg resultGPR = result.gpr();
+        if (character)
+            callOperation(operationStringIndexOfWithIndexWithOneChar, resultGPR, LinkableConstant::globalObject(*this, node), baseGPR, indexGPR, TrustedImm32(character.value()));
+        else
+            callOperation(operationStringIndexOfWithIndex, resultGPR, LinkableConstant::globalObject(*this, node), baseGPR, argumentGPR, indexGPR);
+        exceptionCheck();
+
+        strictInt32Result(resultGPR, node);
+        return;
+    }
+
+    SpeculateCellOperand base(this, node->child1());
+    SpeculateCellOperand argument(this, node->child2());
+
+    GPRReg baseGPR = base.gpr();
+    GPRReg argumentGPR = argument.gpr();
+
+    std::optional<UChar> character;
+    String searchString = node->child2()->tryGetString(m_graph);
+    if (!!searchString) {
+        if (searchString.length() == 1)
+            character = searchString.characterAt(0);
+    }
+
+    speculateString(node->child1(), baseGPR);
+    speculateString(node->child2(), argumentGPR);
```

Comment:
This looks identical to the above.  You can move this up before the `if (node->child3())` check.  The `character` and `searchString` blob can be moved to the top of the function before any of the `SpeculateCellOperand` declarations.

Reply:
I don't think this is better. It rather makes the code hard to read & modify. So rather, we would like to keep them separate. Unifying here does not offer any value, or rather, it introduces more bugs (ArrayIndexOf introduced many issues due to that).